src/core/thread/address_resolver.cpp

```diff
@@ -273,13 +330,15 @@ otError AddressResolver::AddCacheEntry(const Ip6::Address &aEid, Mac::ShortAddre
 
     VerifyOrExit(entry != NULL, error = OT_ERROR_NO_BUFS);
 
-    entry->mTarget   = aEid;
-    entry->mRloc16   = aRloc16;
-    entry->mTimeout  = 0;
-    entry->mFailures = 0;
-    entry->mState    = Cache::kStateCached;
+    entry->mTarget              = aEid;
+    entry->mLastTransactionTime = static_cast<uint32_t>(kLastTransactionTimeInvalid);
+    entry->mRloc16              = aRloc16;
+    entry->mRetryTimeout        = 0;
+    entry->mTimeout             = 0;
+    entry->mFailures            = 0;
+    entry->mState               = Cache::kStateCached;
 
-    MarkCacheEntryAsUsed(*entry);
+    MoveCacheEntryBehindCached(*entry);
```

Comment:
In the case where there are no more invalid entries, the "snoop optimization" could have limited value if the device is receiving lots of messages. On the other hand, we wouldn't want the "snoop optimization" to flush entries that are actually in use.

It would add complexity, but one thought is we could have up to X newly-created "snoop optimization" entries. When they are actually used, then we mark them as such. If they are not used after some timeout, then they may be reclaimed. The reason for the timeout is to allow for response delays (especially in the case of responses coming from a sleepy end device).

This can be worked on in a separate PR, but I thought I'd raise the discussion here. Thoughts?

Reply:
"snoop optimization" always uses the invalid entry or the cached entry of oldest age, this PR here (`MoveCacheEntryBehindCached()`) already ensures that the entry newly added by "snoop optimization" has the oldest age.
For cached entry either added by "snoop optimization" or address query, if used (`Resolve()`) for messages initiated by the device, would have `[fresh](https://github.com/openthread/openthread/blob/master/src/core/thread/address_resolver.cpp#L364)` age.
That is, if the entries added by "snoop optimization" are not used, the last entry added this way would always be preferred to be evicted if required, thus it should have limited impact on the cached entries in use, unless in some extreme scenarios, the device would send messages to the dests (up to `[OPENTHREAD_CONFIG_TMF_ADDRESS_CACHE_ENTRIES](https://github.com/openthread/openthread/blob/5d09ce7124be70edc9b635ca9b8670979dbe9835/src/core/config/tmf.h#L45)`) continuously , and all the dests diff from the sources of messages it receives continuously - which I think should be rare to happen.

Besides for performance consideration, it should always be suggested to configure an appropriate value for `[OPENTHREAD_CONFIG_TMF_ADDRESS_CACHE_ENTRIES](https://github.com/openthread/openthread/blob/5d09ce7124be70edc9b635ca9b8670979dbe9835/src/core/config/tmf.h#L45)` according to the use case.

Thoughts?

Reply:
In the case where there are no more invalid entries, the "snoop optimization" only gets to use one entry. Having only one entry can be limiting in certain scenarios. For example, if an FTD receives a number of request messages from different sources for its children, only the response to one of those sources will get to utilize the "snoop" entry. The rest will need to incur the overhead of Address Query again. Do you agree?

Reply:
As for sizing the number of cache entries, that is difficult to do in the case that you do not know what applications or networks you will be supporting.

Reply:
Agree. 
One question for the `X newly-created "snoop optimization" entries`, is it allocated from heap or some fixed ram space?
If latter, with the special scenario mentioned, we'll have another question regarding the size of `the newly-created "snoop optimization" entries` similar as the size issue of cache entries - we never know the proper value without knowing the use cases, and without proper size, the overhead of AQ still could not be avoided.

To relieve the `only one entry` for "snoop optimization" noted above, we may let the entry newly added this way has age like 'SNOOP_CACHE_AGE'  (= kCachesEntries - X), where X is number of entries which could be used for "snoop optimization", when no available invalid entry.

Thoughts?

Reply:
I agree as well. I think this will be covered with the #4542.

Reply:
Nice discussions an suggestions. I agree and like the idea of allowing some max number of snoop entries and after a new one is added not allowing them to be reclaimed for some timeout interval (giving them a chance to be used).